.github/workflows/main.yml

```diff
@@ -7,30 +7,112 @@ on:
     # run daily, this refreshes the cache
     - cron: '13 2 * * *'
 
+concurrency:  # On new push, cancel old workflows from the same PR, branch or tag:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
+  cancel-in-progress: true
+
 jobs:
   python-test:
     name: Python tests
     runs-on: ubuntu-22.04
     strategy:
       fail-fast: false
       matrix:
-        docker-image:  # See https://hub.docker.com/_/python/tags for images
-          - python:2.7.18-alpine3.11
-          - python:3.11.4-alpine3.18
-    container: ${{ matrix.docker-image }}
+        python-version: ['2.7', '3.11']
```

Comment:
Since all Python files have already been migrated to Python 3, do we still need to continue using Python 2's unittest? If so, there might be some adjustments needed within the unittest.

For instance, Python 2 does not support Chinese characters, but for the sake of code coverage, we need to include Chinese characters in our code. In such cases, Python 2 unittest might raise 'Non-ASCII character' error, and we might need to write some subtle code to work around it.

Reply:
> Since all Python files have already been migrated to Python 3, do we still need to continue using Python 2's unittest?

That depends on whether the product is executing these modules with a python 2 interpreter or not. Until the whole switch is done, I would like to be cautios and keep testing it under python 2

Reply:
I just was to correct @acefei's false claim where he said that Python2 would not support Chinese characters:

That is not true: python2 supports UTF-8 unicode including Chinese characters with it.

The condition is that you may not falsely re-encode strings that are already encoded or re-decode bytes that are already decoded.

That is also true for latin non-ASCII characters found in many European languages as well.

The unit tests of the python-libs repository have extensive unit test where I had the explicit assignment to test the xcp library using Python2 and Python3.

You can look at these tests for reference. The also use multi-width characters like Kanjis and Emojis, one character that is used in a test is a character that contains a full-width world map :-)

If you include literal UTF-8 in Python unit test modules, of course you have to add the encoding header to tell Python2 that literal UTF-8 charcaters ar used in this Python module like these tests do.

I guess this might be the little bit of knowledge that you have been missing and I hope that helps.

Reply:
> That depends on whether the product is executing these modules with a python 2 interpreter or not. Until the whole switch is done, I would like to be cautious and keep testing it under python 2

I see, what you're saying makes sense.


> I just want to correct @acefei's claim where he said that Python2 would not support Chinese characters:

Sorry for the inaccuracy in my statement, and I understand what you're saying.
I actually meant to express that in the default Python 2 encoding, using Chinese characters in Python code would result in an error "non-ASCII", and give an example to illustrate that since the code is already specified to use the Python 3 interpreter, there's no need to include redundant code for Python 2 compatibility in the script.
Anyway, let's keep the testing until the whole switch is done.